dht.go

```diff
@@ -133,33 +149,86 @@ func NewDHTClient(ctx context.Context, h host.Host, dstore ds.Batching) *IpfsDHT
 
 func makeDHT(ctx context.Context, h host.Host, dstore ds.Batching, protocols []protocol.ID) *IpfsDHT {
 	rt := kb.NewRoutingTable(KValue, kb.ConvertPeerID(h.ID()), time.Minute, h.Peerstore())
+	rtRecoveryChan := make(chan *rtRecoveryReq)
 
 	cmgr := h.ConnManager()
+
 	rt.PeerAdded = func(p peer.ID) {
 		cmgr.TagPeer(p, "kbucket", 5)
 	}
+
 	rt.PeerRemoved = func(p peer.ID) {
 		cmgr.UntagPeer(p, "kbucket")
+		go func(rtRecoveryChan chan *rtRecoveryReq) {
+			if rt.Size() == 0 {
+				req := mkRtRecoveryReq()
+				logger.Warningf("rt peer removed notification: RT is empty, will attempt to initiate recovery, reqID=%s", req.id)
+				select {
+				case <-ctx.Done():
+					return
+				case rtRecoveryChan <- req:
+					select {
+					case <-ctx.Done():
+						return
+					case <-req.errorChan:
+						// TODO Do we need to do anything here ?
+					}
+				}
+			}
+		}(rtRecoveryChan)
 	}
 
 	dht := &IpfsDHT{
-		datastore:    dstore,
-		self:         h.ID(),
-		peerstore:    h.Peerstore(),
-		host:         h,
-		strmap:       make(map[peer.ID]*messageSender),
-		ctx:          ctx,
-		providers:    providers.NewProviderManager(ctx, h.ID(), dstore),
-		birth:        time.Now(),
-		routingTable: rt,
-		protocols:    protocols,
+		datastore:      dstore,
+		self:           h.ID(),
+		peerstore:      h.Peerstore(),
+		host:           h,
+		strmap:         make(map[peer.ID]*messageSender),
+		ctx:            ctx,
+		providers:      providers.NewProviderManager(ctx, h.ID(), dstore),
+		birth:          time.Now(),
+		routingTable:   rt,
+		protocols:      protocols,
+		rtRecoveryChan: rtRecoveryChan,
 	}
 
 	dht.ctx = dht.newContextWithLocalTags(ctx)
 
 	return dht
 }
 
+func (dht *IpfsDHT) rtRecovery(proc goprocess.Process) {
+	writeResp := func(errorChan chan error, err error) {
+		select {
+		case <-proc.Closing():
+		case errorChan <- err:
+		}
+		close(errorChan)
+	}
+
+	for {
+		select {
+		case req := <-dht.rtRecoveryChan:
+			if dht.routingTable.Size() == 0 {
```

Comment:
> this check seems redundant, as both processes that send recovery requests down the channel check before sending
> 
> edit: i'm actually a bit confused by this code generally. i get the idea of serializing recovery requests, but this will work its way through them in a pretty tight loop, so i could see multiple recovery request being dispatched without error in a row.

The check is to ensure that if multiple go-routines observe that the RT has become empty & send a request to the channel,  only one requests results in the RT being populated & the remaining become no-op.

Reply:
> in fact, it seems like all this process does is serialize _checks_ of the routing table, but does nothing to actually bootstrap it. it's up to the func that actually creates a `rtRecoveryReq` to _do any work_, and i see a few instances where this is essentially just an info log.

Apologies if this wasn't clear. Yes, we need to add a call to the default seeder implementation in #383 to seed the RT with the 'default bootstrap peers'/'known peers'. There is a TODO for this in the code(line 214) & in the TODO list on this PR.

Reply:
@aarshkshah1992 totally! it really felt like something was missing! definitely missed the comment. with that in mind, the rest are just small comments/docs improvements. generally looking good :)